pwiz/data/vendor_readers/Waters/SpectrumList_Waters.hpp

```diff
@@ -112,6 +115,9 @@ class PWIZ_API_DECL SpectrumList_Waters : public SpectrumListIonMobilityBase
     mutable boost::mutex readMutex;
 
     void createIndex();
+    void createDDAIndex();
+    void getDDAScan(unsigned int index, vector<float>& masses, vector<float>& intensities, IndexEntry& ie) const;

```

Comment:
Unfortunately that IndexEntry reference that gets passed in to this function should have been const. There's no good reason to change the index after it's been generated (the only reason it's mutable is so it can be generated on-demand instead of at construction). This will need to be refactored to return the setMass and precursorMass explicitly (could be floats set by ref or a new struct).

Reply:
> so it can be generated on-demand instead of at construction
Isn't that exactly what we are doing here - generating on demand?

We need other stuff beyond the setMass and precursorMass, we also populate the function and scan; pretty much everything in the index, in fact.

The alternative is to populate the index up-front (now we just set the size up-front); however, it's doing the same work by generating it as we get each scan - I don't see that's actually degrading performance. Or am I missing something?

Reply:
As it happens, at one point it *was* generating up front but Martin who owns the MassLynxSDK decided it would be better to also return the `masses` and `intensities` at the point of getting the scan information, which means we have to generate the index 'on-demand' to avoid getting the same information twice.

If we *really* don't want to generate the index on-demand then we could refactor this again.

Reply:
When a Reader_* is constructed, it also constructs its SpectrumList_* and ChromatogramList_*. Building the index for those can take some time, especially spectra since there are usually more of those. So if the calling code only wants chromatograms, it makes no sense to make them wait for the spectrum list index to be built. So the index is generated on-demand the first time a spectrum, spectrumIdentity, or SpectrumList::size() is called. But that's a vendor-by-vendor thing. I see that in Waters we were NOT doing that because generating the index on construction was never a big delay. Looks like we do it in Agilent:

```
PWIZ_API_DECL size_t SpectrumList_Agilent::size() const
{
    boost::call_once(indexInitialized_.flag, boost::bind(&SpectrumList_Agilent::createIndex, this));
    return size_;
}


PWIZ_API_DECL const SpectrumIdentity& SpectrumList_Agilent::spectrumIdentity(size_t index) const
{
    boost::call_once(indexInitialized_.flag, boost::bind(&SpectrumList_Agilent::createIndex, this));
...snip...
}


PWIZ_API_DECL size_t SpectrumList_Agilent::find(const string& id) const
{
    boost::call_once(indexInitialized_.flag, boost::bind(&SpectrumList_Agilent::createIndex, this));
...snip...
```

The vital thing is the full index must be generated and immutable before the size(), spectrumIdentity(), find(), or spectrum () functions can run. I'm surprised the SpectrumList_Waters test passes because I could of sworn we were testing that spectrumIdentity works in VendorReaderTestHarness. But it looks like I was wrong. Maybe it got buried in a git stash somewhere. :/

Reply:
> The vital thing is the full index must be generated and immutable before the size(), spectrumIdentity(), find(), or spectrum () functions can run

Ah ok, I didn't realise this, I will update the code accordingly